File: Problem-033/solution.js
function digitCancellingFractions() {
  let numeratorproduct = 1;
  let denominatorproduct =1;
  for (let numerator = 10 ; numerator<100; numerator ++){
    for (let denominator = numerator+1; denominator < 100; denominator ++){
      let nstr = numerator.toString();
      let dstr = denominator.toString();
      if (nstr[1] === "0" && dstr[1] === "0"){
        continue;
      }
      for ( let digit of nstr ){
        if ( digit === "0") continue;
        if (dstr.includes(digit)){
          let newN = nstr.replace(digit,"");
          let newD = dstr.replace(digit, "");
          if (newD ==="0") continue;
          if (numerator/denominator === Number(newN)/ Number(newD)){
            numeratorproduct *= numerator;
            denominatorproduct *= denominator;
          }
          break;
        }
      }
    }
  }
  let divisor = gcd(numeratorproduct , denominatorproduct)
  return denominatorproduct/ divisor;
}

function gcd(a,b){
  while(b!==0){
    let temp = b;
    b = a %b;
    a = temp;
  }
  return a;
}
digitCancellingFractions();

File: Problem-033/solution.py
from math import gcd


def digit_cancelling_fractions():
    numerator_product = 1
    denominator_product = 1

    for numerator in range(10, 100):
        for denominator in range(numerator + 1, 100):

            n_str = str(numerator)
            d_str = str(denominator)

            
            if n_str[1] == "0" and d_str[1] == "0":
                continue

            for digit in n_str:
                if digit == "0":
                    continue

                if digit in d_str:

                    new_n = n_str.replace(digit, "", 1)
                    new_d = d_str.replace(digit, "", 1)

                    if new_d == "0":
                        continue

                    if numerator / denominator == int(new_n) / int(new_d):
                        numerator_product *= numerator
                        denominator_product *= denominator

                    break

    divisor = gcd(numerator_product, denominator_product)

    return denominator_product // divisor


print(digit_cancelling_fractions())
